packages/cosmic-swingset/lib/ag-solo/chain-cosmos-sdk.js

```diff
@@ -352,25 +310,107 @@ ${chainID} chain does not yet know of address ${myAddr}${adviseProvision(
+        log.error('WebSocket error', e);
+      });
 
-  // TODO: another way to make this cheaper would be to extract the apphash
-  // from the received block, and only check the mailbox if it changes.
-  // That's more coarse than checking for only our own slot, but better than
-  // hitting the rest-server on every single block.
+      // This magic identifier just distinguishes our subscription
+      // from other noise on the Websocket, if there is any.
+      const MAGIC_ID = 13254;
+      ws.addEventListener('open', _ => {
+        // We send a message to subscribe to every
+        // new block header.
+        const obj = {
+          // JSON-RPC version 2.0.
+          jsonrpc: '2.0',
+          id: MAGIC_ID,
+          // We want to subscribe.
+          method: 'subscribe',
+          params: {
+            // Here is the Tendermint event for new blocks.
+            query: "tm.event = 'NewBlockHeader'",
+          },
+        };
+        // Send that message, and wait for the subscription.
+        ws.send(JSON.stringify(obj));
+      });
+      ws.addEventListener('message', ev => {
+        // We received a message.
+        const obj = JSON.parse(ev.data);
+        if (obj.id === MAGIC_ID) {
+          // It matches our subscription, so notify.
+          updater.updateState(obj);
+        }
+      });
+      ws.addEventListener('close', _ => {
+        // The value `undefined` as the resolution of this retry
+        // tells the caller to retry again with a different RPC server.
+        retryPK.resolve(undefined);
+      });
+      // Return an unresolved promise that resolves to `undefined`
+      // when the WebSocket is closed.
+      return retryPK.promise;
+    });
 
-  c.lightClient.on('update', ({ height }) => {
-    log(`new block on ${GCI}, fetching mailbox`);
-    return getMailbox(height)
-      .then(({ outbox, ack }) => {
-        // console.debug('have outbox', outbox, ack);
-        if (outbox) {
-          inbound(GCI, outbox, ack);
+    return notifier;
+  };
+
+  // Begin the block notifier cycle.
+  const blockNotifier = getBlockNotifier();
+
+  /**
+   * This function is entered at most the same number of times
+   * as the blockNotifier announces a new block.
+   *
+   * It then gets the mailbox.  There are no optimisations.
+   *
+   * @param {number=} lastBlockUpdate
+   */
+  const recurseEachNewBlock = (lastBlockUpdate = undefined) => {
+    blockNotifier
+      .getUpdateSince(lastBlockUpdate)
+      .then(({ updateCount, value }) => {
+        if (!value) {
+          throw Error(`${GCI} unexpectedly finished!`);
         }
-      })
-      .catch(e => log.error(`Failed to fetch ${GCI} mailbox:`, e));
-  });
-  c.lightClient.on('close', e => log.error('closed', e));
+        log.debug(`new block on ${GCI}, fetching mailbox`);
+        getMailbox()
+          .then(({ outbox, ack }) => {
+            // console.debug('have outbox', outbox, ack);
+            if (outbox) {
```

Comment:
I think we need to call `inbound()` even if `outbox` is empty: the chain might be acking our previous message, but not sending anything new of its own.

I see that our old code had the `if (outbox)` check too, so either I was confused back then, or I'm confused now (or both). I know `[]` is truthy, which makes me wonder even more what that check was doing. Is it easy for you to doublecheck the chain side emitter of this query data and see what circumstances might cause `outbox` to be falsy?

The other consideration is that we might have been trying to avoid cranking the kernel when we know there was nothing to do, in case it caused activity in what should have been an idle state. I think we have guards against this later on (we call an add-to-mailbox function and it returns a boolean telling us whether we need to crank the kernel or not). I think we should probably lean towards always cranking the kernel, and just make sure that the device remains properly idle if it sees that there are no new messages (and no new ack).

Reply:
> I see that our old code had the `if (outbox)` check too, so either I was confused back then, or I'm confused now (or both). I know `[]` is truthy, which makes me wonder even more what that check was doing. Is it easy for you to doublecheck the chain side emitter of this query data and see what circumstances might cause `outbox` to be falsy?

I can explain: in the cases where the results of `query swingset mailbox` fails, we are considering the query command to have succeeded, but are currently returning `{}`, which fails the `if (outbox)` check.  I will make this more explicit.